contracts/contracts/strategies/balancer/BalancerMetaPoolStrategy.sol

```diff
@@ -0,0 +1,561 @@
+            .ExitPoolRequest(poolAssets, minAmountsOut, userData, false);
+
+        balancerVault.exitPool(
+            balancerPoolId,
+            address(this),
+            /* Payable keyword is required because of the IBalancerVault interface even though
+             * this strategy shall never be receiving native ETH
+             */
+            payable(address(this)),
+            request
+        );
+
+        // STEP 3 - Convert the balancer pool assets to the vault collateral assets and send to the vault
+        // For each of the Balancer pool assets
+        for (uint256 i = 0; i < tokens.length; ++i) {
+            address poolAsset = address(tokens[i]);
+            // Convert the balancer pool asset to the strategy asset
+            address strategyAsset = fromPoolAsset(poolAsset);
+            // Get the balancer pool assets withdraw from the pool plus any that were already in this strategy contract
+            uint256 poolAssetAmount = IERC20(poolAsset).balanceOf(
+                address(this)
+            );
+
+            // Unwrap assets like wstETH and sfrxETH to rebasing assets stETH and frxETH
+            uint256 unwrappedAmount = 0;
+            if (poolAssetAmount > 0) {
+                unwrappedAmount = unwrapPoolAsset(
+                    strategyAsset,
+                    poolAssetAmount
+                );
+            }
+
+            // Transfer the vault collateral assets to the vault
+            if (unwrappedAmount > 0) {
+                IERC20(strategyAsset).safeTransfer(
+                    vaultAddress,
+                    unwrappedAmount
+                );
+                emit Withdrawal(
+                    strategyAsset,
+                    platformAddress,
+                    unwrappedAmount
+                );
+            }
+        }
+    }
+
+    /**
+     * @notice Approves the Balancer Vault to transfer poolAsset counterparts
+     * of all of the supported assets from this strategy. E.g. stETH is a supported
+     * strategy and Balancer Vault gets unlimited approval to transfer wstETH.
+     *
+     * If Balancer pool uses a wrapped version of a supported asset then also approve
+     * unlimited usage of an asset to the contract responsible for wrapping.
+     *
+     * Approve unlimited spending by Balancer Vault and Aura reward pool of the
+     * pool BPT tokens.
+     *
+     * Is only executable by the Governor.
+     */
+    function safeApproveAllTokens()
+        external
+        override
+        onlyGovernor
+        nonReentrant
+    {
+        uint256 assetCount = assetsMapped.length;
+        for (uint256 i = 0; i < assetCount; ++i) {
+            _abstractSetPToken(assetsMapped[i], platformAddress);
+        }
+        _approveBase();
+    }
+
+    // solhint-disable-next-line no-unused-vars
+    function _abstractSetPToken(address _asset, address) internal override {
+        address poolAsset = toPoolAsset(_asset);
+        if (_asset == stETH) {
```

Comment:
Could we just use poolAsset and skip all the if's?

Reply:
This is used by the `checkBalance` functions [here](https://github.com/OriginProtocol/origin-dollar/pull/1697/files#diff-7d59dfab7a512a013f0696d107a97b7b6b0f1e8d656b8baac6642f61e3d84db6R161) and [here](https://github.com/OriginProtocol/origin-dollar/pull/1697/files#diff-7d59dfab7a512a013f0696d107a97b7b6b0f1e8d656b8baac6642f61e3d84db6R208) and both versions (with asset parameter & without) of checkBalance are used in a few places in the [fork test file](https://github.com/OriginProtocol/origin-dollar/pull/1697/files#diff-4ff4ac51088496da8787ba1d1726b09145dc496c6c9a9f9877bf7c47b3fc2522R665)

Reply:
We could change it to: 
```
address poolAsset = _toPoolAsset(_asset);
IERC20(_asset).approve(poolAsset, type(uint256).max);
```

Though in case of WETH we would be doing: 
```
IERC20(weth).approve(weth, type(uint256).max);
```
To mitigate that we could then:
```
address poolAsset = _toPoolAsset(_asset);
if (poolAsset != _asset) {
  IERC20(_asset).approve(poolAsset, type(uint256).max);
}
```
But we have the if's back and a bit worse readability